File: walking_robot_sim/src/tests/old/cmd_vel_pub.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from quadropted_msgs.msg import RobotVelocity
import math

class RobotVelocityHandler(Node):
    def __init__(self):
        super().__init__('robot_velocity_handler')

        self.declare_parameter('verbose', False)
        self.verbose = self.get_parameter('verbose').get_parameter_value().bool_value
        if self.verbose:
            self.get_logger().info(f"Verbose mode: {self.verbose}")

        self.subscription = self.create_subscription(
            Twist, 
            'cmd_vel',
            self.robot_velocity_callback, 
            10)

        self.publisher_ = self.create_publisher(RobotVelocity, 'robot_velocity', 10)
        if self.verbose:
            self.get_logger().info("Node started: RobotVelocityHandler")

        # Переменная для секундомера
        self.motion_start_time = None

    def robot_velocity_callback(self, msg: Twist):
        # Определяем, есть ли ненулевая скорость
        has_velocity = (msg.linear.x != 0 or msg.linear.y != 0 or 
                        msg.linear.z != 0 or 
                        msg.angular.x != 0 or msg.angular.y != 0 or msg.angular.z != 0)

        current_time = self.get_clock().now()

        # Если есть ненулевая скорость и секундомер не запущен — запускаем его
        if has_velocity and self.motion_start_time is None:
            self.motion_start_time = current_time
            if self.verbose:
                self.get_logger().info(f"Motion started at time: {current_time.to_msg()}")

        # Если скорости нет (робот остановился) и секундомер был запущен — останавливаем его
        if not has_velocity and self.motion_start_time is not None:
            elapsed = current_time - self.motion_start_time
            if self.verbose:
                self.get_logger().info(f"Motion stopped at time: {current_time.to_msg()}")
                self.get_logger().info(f"Elapsed motion time: {elapsed.nanoseconds / 1e9:.3f} seconds")
            self.motion_start_time = None
        if self.verbose:
            self.get_logger().info(
                f"Received Twist: linear=({msg.linear.x}, {msg.linear.y}, {msg.linear.z}), "
                f"angular=({msg.angular.x}, {msg.angular.y}, {msg.angular.z})"
            )

        new_msg = RobotVelocity()
        new_msg.robot_id = 1

        new_msg.cmd_vel.linear.x = self.multiply_and_limit(msg.linear.x, 0.035, -1.0, 1.0)
        new_msg.cmd_vel.linear.y = self.multiply_and_limit(msg.linear.y, 0.012, -1.0, 1.0)
        new_msg.cmd_vel.linear.z = msg.linear.z

        new_msg.cmd_vel.angular.x = msg.angular.x
        new_msg.cmd_vel.angular.y = msg.angular.y
        new_msg.cmd_vel.angular.z = self.limit(msg.angular.z, -1.0, 1.0)

        self.publisher_.publish(new_msg)
        if self.verbose:
            self.get_logger().info(
                f"Published RobotVelocity: robot_id={new_msg.robot_id}, "
                f"linear=({new_msg.cmd_vel.linear.x}, {new_msg.cmd_vel.linear.y}, {new_msg.cmd_vel.linear.z}), "
                f"angular=({new_msg.cmd_vel.angular.x}, {new_msg.cmd_vel.angular.y}, {new_msg.cmd_vel.angular.z})"
            )
        
    def multiply_and_limit(self, value, scale_factor, min_limit, max_limit):
        # Обработка положительных и отрицательных значений отдельно
        if value > 0:
            adjusted_value = value * 0.035
            scaled_value = scale_factor * (1 - math.exp(-100 * adjusted_value))
        else:
            # Для отрицательных значений значение умножаем на -0.035
            adjusted_value = (-value) * 0.035
            scaled_value = -scale_factor * (1 - math.exp(-100 * adjusted_value))
        
        return self.limit_value(scaled_value, min_limit, max_limit)

    def limit_value(self, value, min_limit, max_limit):
        if value > max_limit:
            return max_limit
        elif value < min_limit:
            return min_limit
        else:
            return value

    def limit(self, value, min_limit, max_limit):
        if value > max_limit:
            return max_limit
        elif value < min_limit:
            return min_limit
        else:
            return value

def main(args=None):
    rclpy.init(args=args)
    node = RobotVelocityHandler()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: walking_robot_sim/src/quadropted_controller_cpp/src/nodes/cmd_vel_pub.cpp
#include <cmath>
#include <geometry_msgs/msg/twist.hpp>
#include <quadropted_msgs/msg/robot_velocity.hpp>
#include <rclcpp/rclcpp.hpp>

namespace quadropted {

class CmdVelPub : public rclcpp::Node {
  public:
    CmdVelPub() : Node("cmd_vel_pub_cpp"), motion_start_time_(0.0) {
        declare_parameter("verbose", false);
        verbose_ = get_parameter("verbose").as_bool();

        cmd_vel_sub_ = create_subscription<geometry_msgs::msg::Twist>(
            "cmd_vel", 10, std::bind(&CmdVelPub::twist_callback, this, std::placeholders::_1));

        robot_vel_pub_ = create_publisher<quadropted_msgs::msg::RobotVelocity>("robot_velocity", 10);

        RCLCPP_INFO(get_logger(), "CmdVelPub C++ Node started");
    }

  private:
    void twist_callback(const geometry_msgs::msg::Twist::SharedPtr msg) {
        bool has_velocity =
            (std::abs(msg->linear.x) > 1e-6 || std::abs(msg->linear.y) > 1e-6 || std::abs(msg->linear.z) > 1e-6 ||
             std::abs(msg->angular.x) > 1e-6 || std::abs(msg->angular.y) > 1e-6 || std::abs(msg->angular.z) > 1e-6);

        auto current_time = this->now();
        if (has_velocity && motion_start_time_ == 0.0) {
            motion_start_time_ = current_time.seconds();
            if (verbose_) RCLCPP_INFO(get_logger(), "Motion started at: %.3f", motion_start_time_);
        }

        if (!has_velocity && motion_start_time_ > 0.0) {
            double elapsed = current_time.seconds() - motion_start_time_;
            if (verbose_) RCLCPP_INFO(get_logger(), "Motion stopped, elapsed: %.3f sec", elapsed);
            motion_start_time_ = 0.0;
        }

        auto out = std::make_unique<quadropted_msgs::msg::RobotVelocity>();
        out->robot_id = 1;

        out->cmd_vel.linear.x = multiply_and_limit(msg->linear.x);
        out->cmd_vel.linear.y = multiply_and_limit_y(msg->linear.y);
        out->cmd_vel.linear.z = msg->linear.z;
        out->cmd_vel.angular.x = msg->angular.x;
        out->cmd_vel.angular.y = msg->angular.y;
        out->cmd_vel.angular.z = limit(msg->angular.z);

        robot_vel_pub_->publish(std::move(out));
    }

    double multiply_and_limit(double value) {
        double adjusted = value * 0.035;
        double scaled = 0.035 * (1.0 - std::exp(-100.0 * std::abs(adjusted)));
        return limit(value >= 0 ? scaled : -scaled);
    }

    double multiply_and_limit_y(double value) {
        double adjusted = value * 0.012;
        double scaled = 0.012 * (1.0 - std::exp(-100.0 * std::abs(adjusted)));
        return limit(value >= 0 ? scaled : -scaled);
    }

    static double limit(double value) {
        if (value > 1.0) return 1.0;
        if (value < -1.0) return -1.0;
        return value;
    }

    bool verbose_;
    double motion_start_time_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_sub_;
    rclcpp::Publisher<quadropted_msgs::msg::RobotVelocity>::SharedPtr robot_vel_pub_;
};

}  // namespace quadropted

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<quadropted::CmdVelPub>());
    rclcpp::shutdown();
    return 0;
}
